From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 12 Jan 2025 10:53:42 +0800
Subject: [PATCH] Teleport async if entity was moving to another region at once


diff --git a/src/main/java/me/earthme/luminol/config/modules/fixes/FoliaEntityMovingFixConfig.java b/src/main/java/me/earthme/luminol/config/modules/fixes/FoliaEntityMovingFixConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..0e51d465db3554ac80d00c6b85cc1f013f3949ba
--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/modules/fixes/FoliaEntityMovingFixConfig.java
@@ -0,0 +1,28 @@
+package me.earthme.luminol.config.modules.fixes;
+
+import me.earthme.luminol.config.ConfigInfo;
+import me.earthme.luminol.config.EnumConfigCategory;
+import me.earthme.luminol.config.IConfigModule;
+
+public class FoliaEntityMovingFixConfig implements IConfigModule {
+    @ConfigInfo(baseName = "enabled", comments =
+            """
+                    A simple fix of a issue on folia\s
+                    (Some times the entity would\s
+                    have a large moment that cross the\s
+                    different tick regions and it would\s
+                    make the server crashed)  but sometimes it might doesn't work""")
+    public static boolean enabled = false;
+    @ConfigInfo(baseName = "warn_on_detected")
+    public static boolean warnOnDetected = true;
+
+    @Override
+    public EnumConfigCategory getCategory() {
+        return EnumConfigCategory.FIXES;
+    }
+
+    @Override
+    public String getBaseName() {
+        return "folia.fix_high_velocity_issue";
+    }
+}
