From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 31 Jan 2025 12:54:45 +0800
Subject: [PATCH] Merge paper #12047


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index ba8b5a0ebe652bfaf5c1498c19d12a91a192bf8e..8b661a11ed798186335b4ea6b907eb719cbbf722 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -40,10 +40,10 @@ public final class ActivationRange {
     }
 
     static Activity[] VILLAGER_PANIC_IMMUNITIES = {
-        Activity.HIDE,
-        Activity.PRE_RAID,
-        Activity.RAID,
-        Activity.PANIC
+            Activity.HIDE,
+            Activity.PRE_RAID,
+            Activity.RAID,
+            Activity.PANIC
     };
 
     private static int checkInactiveWakeup(final Entity entity) {
@@ -85,28 +85,28 @@ public final class ActivationRange {
      * @return boolean If it should always tick.
      */
     public static boolean initializeEntityActivationState(final Entity entity, final SpigotWorldConfig config) {
-        return (entity.activationType == ActivationType.MISC && config.miscActivationRange == 0)
-            || (entity.activationType == ActivationType.RAIDER && config.raiderActivationRange == 0)
-            || (entity.activationType == ActivationType.ANIMAL && config.animalActivationRange == 0)
-            || (entity.activationType == ActivationType.MONSTER && config.monsterActivationRange == 0)
-            || (entity.activationType == ActivationType.VILLAGER && config.villagerActivationRange <= 0)
-            || (entity.activationType == ActivationType.WATER && config.waterActivationRange <= 0)
-            || (entity.activationType == ActivationType.FLYING_MONSTER && config.flyingMonsterActivationRange <= 0)
-            || entity instanceof EyeOfEnder
-            || entity instanceof Player
-            || entity instanceof ThrowableProjectile
-            || entity instanceof EnderDragon
-            || entity instanceof EnderDragonPart
-            || entity instanceof WitherBoss
-            || entity instanceof AbstractHurtingProjectile
-            || entity instanceof LightningBolt
-            || entity instanceof PrimedTnt
-            || entity instanceof net.minecraft.world.entity.item.FallingBlockEntity
-            || entity instanceof net.minecraft.world.entity.vehicle.AbstractMinecart
-            || entity instanceof net.minecraft.world.entity.vehicle.AbstractBoat
-            || entity instanceof EndCrystal
-            || entity instanceof FireworkRocketEntity
-            || entity instanceof ThrownTrident;
+        return (entity.activationType == ActivationType.MISC && config.miscActivationRange <= 0) // Luminol - Merge paper #12047
+                || (entity.activationType == ActivationType.RAIDER && config.raiderActivationRange <= 0) // Luminol - Merge paper #12047
+                || (entity.activationType == ActivationType.ANIMAL && config.animalActivationRange <= 0) // Luminol - Merge paper #12047
+                || (entity.activationType == ActivationType.MONSTER && config.monsterActivationRange <= 0) // Luminol - Merge paper #12047
+                || (entity.activationType == ActivationType.VILLAGER && config.villagerActivationRange <= 0)
+                || (entity.activationType == ActivationType.WATER && config.waterActivationRange <= 0)
+                || (entity.activationType == ActivationType.FLYING_MONSTER && config.flyingMonsterActivationRange <= 0)
+                || entity instanceof EyeOfEnder
+                || entity instanceof Player
+                || entity instanceof ThrowableProjectile
+                || entity instanceof EnderDragon
+                || entity instanceof EnderDragonPart
+                || entity instanceof WitherBoss
+                || entity instanceof AbstractHurtingProjectile
+                || entity instanceof LightningBolt
+                || entity instanceof PrimedTnt
+                || entity instanceof net.minecraft.world.entity.item.FallingBlockEntity
+                || entity instanceof net.minecraft.world.entity.vehicle.AbstractMinecart
+                || entity instanceof net.minecraft.world.entity.vehicle.AbstractBoat
+                || entity instanceof EndCrystal
+                || entity instanceof FireworkRocketEntity
+                || entity instanceof ThrownTrident;
     }
 
     /**
@@ -234,8 +234,8 @@ public final class ActivationRange {
             if (entity instanceof final Bee bee) {
                 final BlockPos movingTarget = bee.getMovingTarget();
                 if (bee.isAngry() ||
-                    (bee.getHivePos() != null && bee.getHivePos().equals(movingTarget)) ||
-                    (bee.getSavedFlowerPos() != null && bee.getSavedFlowerPos().equals(movingTarget))
+                        (bee.getHivePos() != null && bee.getHivePos().equals(movingTarget)) ||
+                        (bee.getSavedFlowerPos() != null && bee.getSavedFlowerPos().equals(movingTarget))
                 ) {
                     return 20;
                 }
