From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sat, 30 Nov 2024 12:31:48 +0800
Subject: [PATCH] Threaded region start tick and finished tick event


diff --git a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 056fb1ca7b07d5e713dcbd951830b14fc9025f4c..b67a76347f06686d684215c89259e764f8dd1604 100644
--- a/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -401,7 +401,7 @@ public final class TickRegionScheduler {
 
             final long scheduledStart = this.getScheduledStart();
             final long scheduledEnd = scheduledStart + TIME_BETWEEN_TICKS;
-
+            new me.earthme.luminol.api.events.TickRegionStartTickEvent(this.region == null ? -1 : this.region.region.id,System.nanoTime()); //Luminol - Threaded regions API
             synchronized (this) {
                 this.currentTickData = new TickTime(
                     lastTickStart, scheduledStart, tickStart, cpuStart,
@@ -436,6 +436,7 @@ public final class TickRegionScheduler {
                 );
 
                 this.addTickTime(time);
+                new me.earthme.luminol.api.events.TickRegionFinishedTickEvent(this.region == null ? -1 : this.region.region.id,System.nanoTime()); //Luminol - Threaded regions API
                 TickRegionScheduler.setTickTask(null);
                 if (this.region != null) {
                     TickRegionScheduler.setTickingRegion(null);
